File: src/services/tools/builtin/skills.ts
import { ToolResult } from '../../../types/tools';

// Lazy import to avoid circular dependency issues
let _skillRegistry: typeof import('../../skills').skillRegistry | null = null;

async function getSkillRegistry() {
  if (!_skillRegistry) {
    const { skillRegistry } = await import('../../skills');
    _skillRegistry = skillRegistry;
  }
  return _skillRegistry;
}

export async function executeSkillTools(
  toolName: string,
  input: Record<string, unknown>
): Promise<ToolResult> {
  if (toolName !== 'use_skill') {
    return {
      tool_use_id: '',
      content: `Unknown skill tool: ${toolName}`,
      is_error: true,
    };
  }

  const skillName = input.name as string;
  if (!skillName) {
    return {
      tool_use_id: '',
      content: 'Missing required parameter: name',
      is_error: true,
    };
  }

  const skillRegistry = await getSkillRegistry();
  const instructions = skillRegistry.getSkillInstructions(skillName);
  if (!instructions) {
    const availableSkills = skillRegistry.getSkills().map(s => s.name);
    return {
      tool_use_id: '',
      content: `Unknown skill: ${skillName}. Available skills: ${availableSkills.join(', ')}`,
      is_error: true,
    };
  }

  return {
    tool_use_id: '',
    content: `# Skill: ${skillName}\n\nFollow these instructions to complete the task:\n\n${instructions}`,
  };
}

File: alloy-server/src/tools/skills.rs
//! `use_skill` tool. Returns the named skill's instructions, formatted the
//! same way as the SPA-side executor.

use serde_json::Value;
use std::sync::Arc;

use crate::skill_registry::SkillRegistry;
use crate::tools::input_string;

pub async fn execute(registry: &Arc<SkillRegistry>, input: &Value) -> Result<String, String> {
    let name = input_string(input, "name").unwrap_or("").trim();
    if name.is_empty() {
        return Err("Missing required parameter: name".into());
    }
    match registry.instructions(name) {
        Some(instructions) => Ok(format_skill(name, &instructions)),
        None => {
            let available = registry.available().join(", ");
            Err(format!(
                "Unknown skill: {}. Available skills: {}",
                name, available
            ))
        }
    }
}

/// Format a skill's instructions the way the model should receive them. Shared
/// by the `use_skill` tool and explicit `/skill_name` slash-command invocation.
pub fn format_skill(name: &str, instructions: &str) -> String {
    format!(
        "# Skill: {}\n\nFollow these instructions to complete the task:\n\n{}",
        name, instructions
    )
}

/// The formatted instruction block for a known skill, or `None` if unknown.
pub fn skill_block(registry: &SkillRegistry, name: &str) -> Option<String> {
    registry.instructions(name).map(|i| format_skill(name, &i))
}
